src/main/java/com/google/firebase/FirebaseApp.java

```diff
@@ -446,14 +465,70 @@ protected void scheduleNext(Callable<Void> task, long delayMillis) {
       logger.debug("Scheduling next token refresh in {} milliseconds", delayMillis);
       try {
         future = firebaseApp.schedule(task, delayMillis);
-      } catch (UnsupportedOperationException ignored) {
+      } catch (UnsupportedOperationException e) {
         // Cannot support task scheduling in the current runtime.
+        logger.debug("Failed to schedule token refresh event", e);
+      }
+    }
+
+    /**
+     * Starts the TokenRefresher if not already started. Starts listening to credentials changed
+     * events, and schedules refresh events every time the OAuth2 token changes. If no active
+     * token is present, or if the available token is set to expire soon, this will also schedule
+     * a refresh event to be executed immediately.
+     *
+     * <p>This operation is idempotent. Calling it multiple times, or calling it after the
+     * refresher has been stopped has no effect.
+     */
+    final synchronized void start() {
+      // Allow starting only from the ready state.
+      if (!state.compareAndSet(STATE_READY, STATE_STARTED)) {
+        return;
+      }
+
+      logger.debug("Starting the proactive token refresher");
+      credentials.addChangeListener(this);
+      AccessToken accessToken = credentials.getAccessToken();
+      long refreshDelay = 0L;
+      if (accessToken != null) {
+        refreshDelay = Math.max(getRefreshDelay(accessToken), refreshDelay);
       }
+      // If the access token is null, or is about to expire (i.e. expires in less than 5 minutes),
+      // schedule a refresh event with 0 delay. Otherwise schedule a refresh event at the token
+      // expiry time, minus 5 minutes.
+      scheduleRefresh(refreshDelay);
```

Comment:
Subtract the 5 minutes here

Reply:
This happens in the call to `getRefreshDelay()` a couple of lines earlier. I reorg'ed the code and comments a bit to make this clear.

Reply:
If you substract the 5 minutes here, then getRefreshDelay() will return exactly the delay that is specified in the token and then "refreshDelay > 0" check will work for all delays.